sdk/python/kfp/components/base_model.py

```diff
@@ -0,0 +1,424 @@
+# Copyright 2022 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import dataclasses
+import inspect
+import json
+import pprint
+from collections import abc
+from typing import (Any, Dict, ForwardRef, Iterable, Iterator, Mapping,
+                    MutableMapping, MutableSequence, Optional, OrderedDict,
+                    Sequence, Tuple, Type, TypeVar, Union)
+
+PRIMITIVE_TYPES = {int, str, float, bool}
+
+# typing.Optional.__origin__ is typing.Union
+UNION_TYPES = {Union}
+
+# do not need typing.List, because __origin__ is list
+ITERABLE_TYPES = {
+    list,
+    abc.Sequence,
+    abc.MutableSequence,
+    Sequence,
+    MutableSequence,
+    Iterable,
+}
+
+# do not need typing.Dict, because __origin__ is dict
+MAPPING_TYPES = {
+    dict, abc.Mapping, abc.MutableMapping, Mapping, MutableMapping, OrderedDict,
+    collections.OrderedDict
+}
+
+OTHER_SUPPORTED_TYPES = {type(None), Any}
+SUPPORTED_TYPES = PRIMITIVE_TYPES | UNION_TYPES | ITERABLE_TYPES | MAPPING_TYPES | OTHER_SUPPORTED_TYPES
+
+BaseModelType = TypeVar('BaseModelType', bound='BaseModel')
+
+
+class BaseModel:
```

Comment:
I could be wrong, but my sense is we may not need a `BaseModel` class at all. Because IR is defined via protobuf, we got the serialization/deserialization functionality via the protobuf library already, any other serialization/deserialization logic seems to me redundant and reimplementing the wheel of protobuf library? Another way to think is what would be the use case for `from_json` and `to_json` methods in this base model? 

We need some logic to map between our existing internal structures and protobuf generated structures, I suspect the conversion logic would be customized per each individual class, thus a base class may not add much value to the conversion.

Reply:
A base class could be useful if our classes and the protobuf generated classes have exact same fields, and the base class could do some `setattr`/`getattr` via field names. That being said, I'm not sure if this is worth it, the downside could be less readable, error-prone, and hard for debugging.

Reply:
Thanks, @chensun. I definitely agree with the sentiment, though a few sources of complexity arose during my development that lead me to take this approach. I'm curious what you think (1 and 2 are cases for using a `BaseModel`, 3 and 4 are other relevant notes):
1. As you mention, we need somewhere to hold our data structures in memory before converting to a `BaseModel`. In the simplest form, I think this would probably be an `@dataclasses.dataclass`. This `BaseModel` accomplishes this: it turns subclasses into `dataclass`es, but does so via inheritance (as opposed to decoration), so we can add extra functionality as needed. This functionality currently includes:
- support for custom `validate_*` methods for more helpful use error messages and `transform_*` methods for data conversion functionality not supported by pure dataclasses
- assertion that the types provided for the dataclass fields are supported by our downstream logic

2. To answer your question:
> what would be the use case for `from_json` and `to_json` methods in this base model?

The main use case (at the end state of _all_ of the V2 YAML component packaging format changes) is enabling support for V1 and V2 readable YAML (read) backward compatibility. This requires at least `from_json`, I think. We want to be able to read in old YAML to the internal structures. The custom implementation of `from_json` (and under the hood, `from_dict`) handles our specific type casting logic, as well as built-in support for aliasing, via the `by_alias` parameter.

3. I began down the route of jointly removing `pydantic` and manipulating our internal-structure-to-proto conversion logic and found this very error prone and challenging to debug. This is because `pydantic` is all or nothing -- there's no way to remove/replace functionality piecewise (structure by structure or method by method) to verify tests are passing throughout the (large) development process. Migrating to a `BaseClass` that we own/control as a discrete step before implementing the IR conversion process alleviates this, reducing the complexity and increasing the observability of the `to_proto` and `from_proto` implementations that come next.

4. Regarding your comment:
> I suspect the conversion logic would be customized per each individual class

I completely agree with this and, if it weren't for the other points above, I would also support not using a `BaseModel` class on this basis.

Also, as a minor point: I'm not yet sure about this, but I think it might also be possible that we'd want the `BaseModel` to own the `to_proto` logic, then the subclasses to own their own encoding handlers (akin to the `cls` argument in [json.dump](https://docs.python.org/3/library/json.html#json.dump). This may not be the direction we ultimately go, but just a thought.

-- 
I suspect there are several things I'm not thinking about here, so curious to hear what you think and talk through it.